### ISSUE TITLE
Implement generic interface for pushing various datatypes to various destinations

### ISSUE THREAD

**Issue by cdoern** [instructlab/instructlab#2672]
Instructlab should have an `Uploader`, `Push`, or, `Publish` class/interface which is expandable for different types (datasets, models, etc). and different destinations (HF, S3, OCI). Having this functionality generalized in a very pluggable way will allow us to easily add new endpoints as we see fit.

**Comment by mbestavros** [comment]
Interested in helping with this!

**Comment by nathan-weinberg** [comment]
I've introduced a `ModelUploader` class as part of https://github.com/instructlab/instructlab/pull/2768 - this is an [Abstract Base Class](https://docs.python.org/3/library/abc.html) that all is used as a parent for all the child uploader classes (e.g. `HFModelUploader`, `OCIModelUploader`, etc.) 

I think we could further abstract this to account for other types of data we want to upload - i.e. an `Uploader` class that could serve as a parent for `ModelUploader` as well as other datatype classes e.g. `DataUploader` 

@cdoern @mbestavros wdyt?

**Comment by cdoern** [comment]
I think the class you introduced makes sense @nathan-weinberg . The further abstraction would be nice, but let's table that discussion until we get to data management. Should we consider this done?

**Comment by nathan-weinberg** [comment]
If @mbestavros is in agreement, we could consider this done - @ktam3 what is the acceptance criteria for this?

**Comment by mbestavros** [comment]
Agreed with @cdoern - data management will probably need some flexibility in the concepts and inputs, but that's a bridge we can cross later. I'm cool calling this done.

**Comment by nathan-weinberg** [comment]
If Kim is fine with it then I say we close this out, as well as the related Jira.

**Comment by ktam3** [comment]
we can close this
